Quick & dirty support for Rails & ActiveRecord 7.1

taylorthurlow opened the pull request:
This adds Ruby 3.1, 3.2, and AR 7.1 to the test matrix with appropriate exclusions, and then makes some tweaks to support ActiveRecord 7.1. I don't have much familiarity with any of the ActiveRecord code, nor do I have any experience with nulldb's code, but I'm putting this PR here in case it may be helpful. I would consider the work here to be pretty sloppy - basically I was just trying to get all the tests to pass. I obviously won't be offended if all of this is scrapped for a more well thought out upgrade path.

Some random notes:
- The `#insert` method in `core.rb` seems to want to return an array of object IDs instead of single object IDs now, couldn't see where in the ActiveRecord codebase this change actually happens, but it seems to be fine the way I have changed it - it only caused problems on Ruby 2.7 and 3.0 when it was not returning an Array. No idea why that is the case.
- AR >= 7.1 perform column creation option checks which means some tests for nonstandard column types were failing because they have unrecognized option names. I monkey patched that check out for AR >= 7.1.
- Other fixes were just method signature problems.

Tests pass in my fork so they should pass here if the workflow run is approved.

tsmartt commented:
Thanks for this. Can you also update https://github.com/taylorthurlow/nulldb/blob/master/activerecord-nulldb-adapter.gemspec#L23

to account for active record 7.1?

ghiculescu commented on lib/active_record/connection_adapters/nulldb_adapter/core.rb:
```diff
@@ -164,22 +164,28 @@ def execute(statement, name = nil)
       end
     end
+
+    Array.wrap(result)
```
Saw this PR after doing https://github.com/nulldb/nulldb/pull/125

I think for this method we should only return an array if the `returning` prop is provided. This matches how other adapters work.

taylorthurlow replied:
Thanks for the extra insight, fixed this.

taylorthurlow commented:
Fixed the gemspec version restriction and implemented @ghiculescu's suggested fix for the array wrapping problem. Feel free to move forward with either of the two PRs, in the end we seem to have done largely the same thing.

gstokkink commented:
@tsmartt can we get this released? Thank you in advance!

cabello commented:
v1.0 is now live: https://rubygems.org/gems/activerecord-nulldb-adapter/versions/1.0.0
